### PR TITLE
ru_obscenity_classifier model have been created

### PR DESCRIPTION
**Pull request by SultanovAR**
ru_obscenity_classifier it is model that classifies text for obscenity, for russian language
Code is placed in `deeppavlov/models/classifiers/ru_obscenity_classifier.py`
Config is placed in `deeppavlov/configs/classifiers/ru_obscenity_classifier.json`

example of working:

```
python deep.py interact -d configs/classifiers/ru_obscenity_classifier.json
2019-06-13 19:12:23.45 INFO in 'deeppavlov.core.data.utils'['utils'] at line 63: Downloading from http://files.deeppavlov.ai/models/obscenity_classifier/ru_obscenity_dataset.zip?config=ru_obscenity_classifier to /home/azat/.deeppavlov/downloads/ru_obscenity_dataset.zip
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17.7k/17.7k [00:00<00:00, 32.2MB/s]
2019-06-13 19:12:23.49 INFO in 'deeppavlov.core.data.utils'['utils'] at line 201: Extracting /home/azat/.deeppavlov/downloads/ru_obscenity_dataset.zip archive into /home/azat/.deeppavlov/downloads/obscenity_dataset
[nltk_data] Downloading package punkt to /home/azat/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /home/azat/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package perluniprops to
[nltk_data]     /home/azat/nltk_data...
[nltk_data]   Package perluniprops is already up-to-date!
[nltk_data] Downloading package nonbreaking_prefixes to
[nltk_data]     /home/azat/nltk_data...
[nltk_data]   Package nonbreaking_prefixes is already up-to-date!
2019-06-13 19:12:24.502 INFO in 'deeppavlov.models.classifiers.ru_obscenity_classifier'['ru_obscenity_classifier'] at line 77: Initializing `RuObscenityClassifier`
text::Ты милый
>> not_obscene
text::Ты сука
>> obscene
text::
```

### CODE REVIEW

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+from typing import List, Union
+from pathlib import Path
+import json
+import pymorphy2
+import re
+from logging import getLogger
+import os
+import errno
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]: PEP 8 imports order

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            r"(\b(?!вело|ски|эн)\w*[пpp][eеиi][дd][oaоаеeирp](?![цянгюсмйчв])[рp]?(?![лт])\w*\b)",
+            r"(\b(?!в?[ст]{1,2}еб)(?:(?:в?[сcз3о][тяaа]?[ьъ]?|вы|п[рp][иоo]|[уy]|р[aа][з3z][ьъ]?|к[оo]н[оo])?[её]б[а-яa-z]*)|(?:[а-яa-z]*[^хлрдв][еeё]б)\b)",
+            r"(\b[з3z][аaоo]л[уy]п[аaeеин]\w*\b)",
+        ))
+
+        return PATTERN_1, PATTERN_2
+
+    def __init__(self, data_path: Union[Path, str], *args, **kwargs):
+        log.info(f"Initializing `{self.__class__.__name__}`")
+        data_path = Path(expand_path(data_path))
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```suggestion
        data_path = expand_path(data_path)
```
`expand_path` already returns a `Path`

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        required_files = ['obscenity_words.json',
+                          'obscenity_words_exception.json',
+                          'obscenity_words_extended.json']
+        for file in required_files:
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]: You should not name a variable `file`

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            if not (data_path/file).exists():
+                raise FileNotFoundError(errno.ENOENT,
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]: Wouldn't this error get raised anyway at the moment of reading the data?

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                                        os.strerror(errno.ENOENT),
+                                        str(data_path/file))
+
+        self.obscenity_words = set(json.load(
+            open(data_path/'obscenity_words.json', encoding="utf-8")))
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```python3
with open(data_path/'obscenity_words.json', encoding="utf-8") as f:
  self.obscenity_words = set(json.load(f))
```

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            open(data_path/'obscenity_words_exception.json', encoding="utf-8")))
+        self.obscenity_words.update(self.obscenity_words_extended)
+
+        PATTERN_1, PATTERN_2 = self._get_patterns()
+        self.regexp = re.compile(PATTERN_1, re.U | re.I)
+        self.regexp2 = re.compile(PATTERN_2, re.U | re.I)
+        self.morph = pymorphy2.MorphAnalyzer()
+        self.word_pattern = re.compile(r'[А-яЁё]+')
+
+    def _check_obscenity(self, text):
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]: Add typing please

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        """it decides whether text is obscene or not
+
+        Args:
+            texts: list of texts, for which it needs to decide they are obscene or not
+
+        Returns:
+            list of strings: 'obscene' or 'not_obscene'
+        """
+        decisions = list(map(self._check_obscenity, texts))
+        return decisions
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```suggestion
        return decisions

```

**Comment by SultanovAR** [comment]
> Add an interaction test for this, please.
> You can add your config to https://github.com/deepmipt/DeepPavlov/blob/master/tests/test_quick_start.py

Tests have been added

Passing of tests:
```
========================================================================================== test session starts ==========================================================================================
platform linux -- Python 3.6.7, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/azat/anaconda3/envs/obscenity_pull_request/bin/python
cachedir: .pytest_cache
rootdir: /home/azat/Programs/obscenity_request/src/deeppavlov
collected 429 items / 425 deselected / 4 selected                                                                                                                                                       

test_quick_start.py::TestQuickStart::test_interacting_pretrained_model[classifiers-classifiers/ru_obscenity_classifier.json-classifiers-IP] PASSED                                                [ 25%]
test_quick_start.py::TestQuickStart::test_interacting_pretrained_model_api[classifiers-classifiers/ru_obscenity_classifier.json-classifiers-IP] PASSED                                            [ 50%]
test_quick_start.py::TestQuickStart::test_serialization[classifiers-classifiers/ru_obscenity_classifier.json-classifiers-IP] SKIPPED                                                              [ 75%]
test_quick_start.py::TestQuickStart::test_consecutive_training_and_interacting[classifiers-classifiers/ru_obscenity_classifier.json-classifiers-IP] SKIPPED                                       [100%]

=========================================================================================== warnings summary ============================================================================================
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/_pytest/mark/structures.py:337
  /home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/_pytest/mark/structures.py:337: PytestUnknownMarkWarning: Unknown pytest.mark.slow - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    PytestUnknownMarkWarning,

/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/_pytest/mark/structures.py:337
  /home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/_pytest/mark/structures.py:337: PytestUnknownMarkWarning: Unknown pytest.mark.gpu_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    PytestUnknownMarkWarning,

/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/framework/ops.py:923
  /home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/framework/ops.py:923: DeprecationWarning: builtin type EagerTensor has no __module__ attribute
    EagerTensor = c_api.TFE_Py_InitEagerTensor(_EagerTensorBase)

/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
/home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
  /home/azat/anaconda3/envs/obscenity_pull_request/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
    return _inspect.getargspec(target)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=================================================================== 2 passed, 2 skipped, 425 deselected, 31 warnings in 5.87 seconds ====================================================================
```

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        return 'not_obscene'
+
+    def __call__(self, texts: List[str]) -> List[str]:
+        """it decides whether text is obscene or not
+
+        Args:
+            texts: list of texts, for which it needs to decide they are obscene or not
+
+        Returns:
+            list of strings: 'obscene' or 'not_obscene'
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]: Are you sure it's better than `True` and `False`?

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            r"(\b\w*([оo]т|вы|[рp]и|[оo]|и|[уy]){0,1}([пnрp][iиеeё]{0,1}[3zзсcs][дd])\w*\b)",
+            r"(\b(вы)?у?[еeё]?би?ля[дт]?[юоo]?\w*\b)",
+            r"(\b(?!вело|ски|эн)\w*[пpp][eеиi][дd][oaоаеeирp](?![цянгюсмйчв])[рp]?(?![лт])\w*\b)",
+            r"(\b(?!в?[ст]{1,2}еб)(?:(?:в?[сcз3о][тяaа]?[ьъ]?|вы|п[рp][иоo]|[уy]|р[aа][з3z][ьъ]?|к[оo]н[оo])?[её]б[а-яa-z]*)|(?:[а-яa-z]*[^хлрдв][еeё]б)\b)",
+            r"(\b[з3z][аaоo]л[уy]п[аaeеин]\w*\b)",
+        ))
+
+        return PATTERN_1, PATTERN_2
+
+    def __init__(self, data_path: Union[Path, str], *args, **kwargs):
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```suggestion
    def __init__(self, data_path: Union[Path, str], *args, **kwargs) -> None:
```

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+log = getLogger(__name__)
+
+
+@register("ru_obscenity_classifier")
+class RuObscenityClassifier(Component):
+    """Rule-Based model that decides whether the sentence is obscene or not,
+    for russian language
+
+    Args:
+        data_path: a directory where required files are storing
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```suggestion
        data_path: a directory where the required files are stored
```

File: deeppavlov/models/classifiers/ru_obscenity_classifier.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+from deeppavlov.core.models.estimator import Component
+from deeppavlov.core.commands.utils import expand_path
+
+log = getLogger(__name__)
+
+
+@register("ru_obscenity_classifier")
+class RuObscenityClassifier(Component):
+    """Rule-Based model that decides whether the sentence is obscene or not,
+    for russian language
```
Comment (yoptar) [inline deeppavlov/models/classifiers/ru_obscenity_classifier.py]:
```suggestion
    for Russian language
```

**Comment by SultanovAR** [comment]
I have added more information about required files in docstring and  now file `obscenity_words_extended.json` is not obligatory

**Pull request merged**